onnx/defs/generator/defs.cc

```diff
@@ -523,4 +527,185 @@ ONNX_OPERATOR_SET_SCHEMA(
 
+static const char* Range_ver11_doc = R"DOC(
+Generate a tensor containing a sequence of numbers that begin at `start` and extends by increments of `delta` 
+up to `limit` (exclusive).)DOC";
+
+template <typename T>
+inline int compute_output_dim_for_range(
+    const TensorProto* start,
+    const TensorProto* limit,
+    const TensorProto* delta) {
+  const auto& start_data = ParseData<T>(start);
+  const auto& limit_data = ParseData<T>(limit);
+  const auto& delta_data =
+      delta != nullptr ? ParseData<T>(delta) : std::vector<T>(1, 1);
+
+  if (start_data.size() != 1 || limit_data.size() != 1 ||
+      delta_data.size() != 1) {
+    fail_shape_inference(
+        "Input to 'Range' op should be scalars (Tensor with only one element)");
+  }
+
+  int n = static_cast<int>(
+      ceil((1.0 * (limit_data[0] - start_data[0])) / delta_data[0]));
+  if (n < 0)
+    n = 0;
+
+  return n;
+}
+
+const std::vector<NodeProto> build_nodes_range_op() {
+    // body for 'Loop node'
+    GraphProto loop_sub_graph;
+    loop_sub_graph.set_name("loop body attribute");
+
+    // 'Loop' node 'body' attribute's graph inputs
+    auto* input_value_info_proto_0 = loop_sub_graph.add_input();
+    input_value_info_proto_0->set_name("i");
+
+    auto* input_value_info_proto_1 = loop_sub_graph.add_input();
+    input_value_info_proto_1->set_name("cond");
+
+    auto* input_value_info_proto_2 = loop_sub_graph.add_input();
+    input_value_info_proto_2->set_name("prev");
+
+    // 'Loop' node 'body' attribute's graph nodes
+    auto* node_proto_0 = loop_sub_graph.add_node();
+    node_proto_0->set_op_type("Identity");
+    node_proto_0->add_input();
+    node_proto_0->set_input(0, "cond");
+    node_proto_0->add_output();
+    node_proto_0->set_output(0, "cond_out");
+
+    auto* node_proto_1 = loop_sub_graph.add_node();
+    node_proto_1->set_op_type("Add");
+    node_proto_1->add_input();
+    node_proto_1->set_input(0, "prev");
+    node_proto_1->add_input();    
+    node_proto_1->set_input(1, "delta");
+    node_proto_1->add_output();
+    node_proto_1->set_output(0, "current");
+
+    auto* node_proto_2 = loop_sub_graph.add_node();
+    node_proto_2->set_op_type("Identity");
+    node_proto_2->add_input();
+    node_proto_2->set_input(0, "prev");
+    node_proto_2->add_output();
+    node_proto_2->set_output(0, "range");
+
+    // 'Loop' node 'body' attribute's graph inputs
+    auto* output_value_info_proto_0 = loop_sub_graph.add_output();
+    output_value_info_proto_0->set_name("cond_out");
+
+    auto* output_value_info_proto_1 = loop_sub_graph.add_output();
+    output_value_info_proto_1->set_name("current");
+
+    auto* output_value_info_proto_2 = loop_sub_graph.add_output();
+    output_value_info_proto_2->set_name("range");
+
+    return FunctionBodyHelper::BuildNodes({
+      // nodes: {outputs, op, inputs, attributes}
+      {{"sub_result"}, "Sub", {"limit", "start"}},
+      {{"delta_cast"}, "Cast", {"delta"}, {{"to", static_cast<int64_t>(1)}}},
```

Comment:
The only issue is "delta" is an optional input. But I don't know how to deal with optional inputs in the 'Functions' world. Any pointers @linkerzhang please ?

Reply:
We can make it non-optional.

Reply:
That looks a bit ugly as usually delta is 1 and thus would be cleaner to make it optional. If the way to support optional inputs in functions is cumbersome, then we can make it optional. Any thoughts ?

Reply:
For optional inputs, we use `empty string` as their names.

Reply:
I did try that, but the function validator failed. I need to check if the validator has a bug.

Reply:
Now, I think we should use non-optional. Here is the reason.
The invoker of function needs to be aware of optional inputs of functions. If an optional input is empty, the runtime has to look into its spec and generate the right optional value. It means --- runtime will have an implementation for each function with some optional inputs, and therefore breaks the idea of having function.

Reply:
Ok - making it non-optional is ugly, but I see no other choice. making it non-optional.

Reply:
This is essentially a limitation of the function construct. We have been considering extending the function construct so that, instead of registering a fixed constant graph as its body, we register a lambda, similar to the type-and-shape-inference lambda, which takes a "context" (which will give us access to the types of inputs, information about attribute values, and which inputs/outputs are present/required), and returns a subgraph. With such an extension, it should be easy to support optional inputs like delta.

Reply:
Thanks. That would be great. Looking forward to it. 

For now, resolving his comment as there is no action needed.